project/driver_controls/src/direction_state.c

```diff
@@ -0,0 +1,39 @@
+#include "direction_state.h"
+
+FSM_DECLARE_STATE(state_neutral);
+FSM_DECLARE_STATE(state_forward);
+FSM_DECLARE_STATE(state_reverse);
+
+// State machine transition tables
+
+FSM_STATE_TRANSITION(state_neutral) {
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_DRIVE, state_forward);
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_REVERSE, state_reverse);
+}
+
+FSM_STATE_TRANSITION(state_forward) {
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_NEUTRAL, state_neutral);
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_REVERSE, state_reverse);
+}
+
+FSM_STATE_TRANSITION(state_reverse) {
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_DRIVE, state_forward);
+  FSM_ADD_TRANSITION(INPUT_EVENT_DIRECTION_SELECTOR_NEUTRAL, state_neutral);
+}
+
+// State output functions
+static void prv_state_neutral(FSM *fsm, const Event *e, void *context) {
+  *(bool*)fsm->context = !(e->id == INPUT_EVENT_GAS_COAST || e->id == INPUT_EVENT_GAS_PRESSED);
```

Comment:
I hope you've gotten the message about not casting and setting values in the same line. This is really confusing.